### DISCUSSION TITLE
HTTPS Imports

### DISCUSSION THREAD

**Discussion by bmeck** [nodejs/help#4909] [Ideas]
Github PR UI is not well suited to this level of conversation

~~Discussing PR https://github.com/nodejs/node/pull/36328~~

Discussing implications of HTTPS imports as a whole

**Comment by bmeck** [comment]
Concern To Mitigate: IO Timeouts

**Reply by jasnell** [reply]
Closely related to this is a never-ending import (a malicious server sending a never-ending stream of data). 

Likely need a max import *size* as well as timeouts.

**Reply by bmeck** [reply]
@jasnell agreed, but there are multiple concepts of granularity here. Graph size vs individual source text size. I think we have to solve for both either by mitigating or scoping out. I'm thinking for this topic we are really talking about mitigations on the individual level, but the infinite recursion might be better for the strongly connected static graph mitigations. I think at least for the initial approach to size, byte limits might be enough as laid out in https://github.com/bmeck/node/tree/policy-max-body . Timeouts are more complex.

**Comment by bmeck** [comment]
Concern To Mitigate: Infinite Recursive Importing

**Reply by targos** [reply]
How can that happen?

**Reply by bmeck** [reply]
A server could dynamically generate a payload that ~= `import "import.meta.url + `1` "`. This would prevent the ESM graph from ever evaluating any code and memory leak until the process dies.

**Reply by bmeck** [reply]
It is possible to do this in CJS already, but with ESM this is a bit different. CJS can write files as they are loaded. ESM must load files after they are written using dynamic import. With the introduction of Loaders and/or 3rd party actors like servers that makes static import also able to cause infinite recursion.

I'm not entirely sure how we want to limit this, could limit the dep graph source size of an import or the total distance of the graph from entry or from the main entry point. When we get to that point if the problem was through static imports it won't be too much of an issue since we can GC the modules prior to linking (linking makes them live for the life of the global unfortunately [cantfix / upstream]). With dynamic imports, we couldn't reclaim that lost memory.

**Reply by jasnell** [reply]
POC:

```js
const fastify = require('fastify')({ logger: true })

let counter = 1;

fastify.get('/foo.ts', (request, reply) => {
  reply
    .code(200)
    .header('Content-Type', 'application/x-typescript')
    .send(`import * as foo from 'http://localhost:3002/foo.ts?${counter++}'`);
})

const start = async () => {
  try {
    await fastify.listen(3002)
    fastify.log.info(`server listening on ${fastify.server.address().port}`)
  } catch (err) {
    fastify.log.error(err)
    process.exit(1)
  }
}

start();
```

**Reply by bmeck** [reply]
The more I stare at this them more it looks like just placing a config on graphs likely ends up limiting the whole program size. If we just limit statically linked graphs E.G. `import "a";` we don't cover a server sending `import("a")`. If we do cover `import("a")` there isn't a clear boundary I can discern on what the limiting factor is since they could do `import("https://alias/a")` instead.

**Comment by bmeck** [comment]
Concern To Mitigate: Server Downtime

**Reply by bmeck** [reply]
We can somewhat do this already with URL redirects, see a policy with:

```json
{
  "scopes": {
    "file:": {
       "integrity": true,
       "dependencies": {
           "https://example.com/foo": "file://cache/https/example.com/foo"
       }
    }
  }
}
```

That would not cover the HTTP redirect semantics, leak the wrong `import.meta.url`, and would not respect cache / updating. It does bring up some questions though about what the update semantics should be for a cache. Should it ever be automatic? I somewhat lean on "no it should always be a manual step to sync the cache".

**Reply by targos** [reply]
I agree the cache update should be manual. I think we need a good CLI to work with it though.

**Comment by bmeck** [comment]
Concern To Mitigate: Alteration of Source Code

**Reply by bmeck** [reply]
* We already have integrity checks via policies, but the use cases of HTTPS imports also can lean towards automatic updating which wouldn't work with integrity checks.

HTTPS has other means of granting trust such as certificate verification of various types.

* Support for [HPKP](https://en.wikipedia.org/wiki/HTTP_Public_Key_Pinning) and/or [Certificate Transparency](https://en.wikipedia.org/wiki/Certificate_Transparency) would rely on a similar need for out of band persistence as policies but does allow for ensuring that the entity that provided the previous source text is the same as the current source text.
* Root CAs are already configurable but not able to be configured solely for import purposes.

**Reply by mcollina** [reply]
This is really important to provide safety against social engineering attacks, e.g. `left-pad` and `event-stream`.

**Reply by bmeck** [reply]
@mcollina those attacks would already be mitigated by policies as they exist today. Is there a specific aspect of those that you are seeking to guard against and should it be outside of the scope of HTTPS itself?

**Reply by mcollina** [reply]
They are not mitigated if you consider availability attacks across multiple versions of the module, e.g. "left-pad" or "event-stream". as you say:

> the use cases of HTTPS imports also can lean towards automatic updating which wouldn't work with integrity checks.

Note that the lack of a central authority makes this case significantly more severe than npm.

**Reply by bmeck** [reply]
> Note that the lack of a central authority makes this case significantly more severe than npm.

That is both a pro and a con. As we see today with Google, having a single point of failure can be problematic. We need to mitigate these issues still in those cases such as http proxies , localized redirection , localized cache etc.

**Reply by bmeck** [reply]
Related concern: https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610 this attack was due to not having an ability to constrain a dependency to only be installed from a specific source (registry in this case). This is solved on web already by general HTTPS pinning techniques so likely wouldn't be vulnerable if we enforce pinning to use HTTPS by default.

**Reply by mcollina** [reply]
> Related concern: https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610 this attack was due to not having an ability to constrain a dependency to only be installed from a specific source (registry in this case). This is solved on web already by general HTTPS pinning techniques so likely wouldn't be vulnerable if we enforce pinning to use HTTPS by default.

We are not in agreement. Domain name ownership changes - one dev might just get the domain name he is using for some really popular library (left-pad) expire. This makes URLs are even more vulnerable to the same attack because there is no real way to differentiate between a legit and illegit domains by just looking at the URL. They both are legit - just at different point in time. A change of ownership can be extremely bad and signal an attack.

It's possible to work around this problem by storing the TLS certificate and validating that the content is always downloaded from the same certificate. At the end, a user should be in charge of vetting/validating if the SSL certificate has changed and if it's still a valid one.

(I generally prefer a provenance solution on the module itself - i.e. signed modules)

**Reply by bmeck** [reply]
@mcollina

> (I generally prefer a provenance solution on the module itself - i.e. signed modules)

How is this different from:

> It's possible to work around this problem by storing the TLS certificate and validating that the content is always downloaded from the same certificate. At the end, a user should be in charge of vetting/validating if the SSL certificate has changed and if it's still a valid one. 

In general code signing is just a temporary and potentially spoofed over time validation of who authored something not that the author is trusted. You get into things like timing servers to start other things that need to be validated consistently over time just like HTTPS.

**Reply by mcollina** [reply]
let's imagine a workflow:

1. I start using a new dependency, I quickly take a look at it and it fits my need, therefore I trust this dependency and its author
2. I configure my "update mechanisms" to pick up any new version of that library authored by the same author
3. If/when the author changes, I'm getting notified to look at that again

This workflow works the same for both certificate pinning and code signing at the module level. I prefer the module level because it's out-of-band of the transport mechanism and it can be easily proxied through. Using TLS certs for this would mean that proxy would need to do something special to keep provenance.

**Reply by bmeck** [reply]
I think we disagree then.

> I configure my "update mechanisms" to pick up any new version of that library authored by the same author

I don't think trust based upon social audits of specific maintainers is either practical or even reliable.

1. We have examples of actors publishing on behalf of others like https://eslint.org/blog/2018/07/postmortem-for-malicious-package-publishes . This is a general workflow for most libraries with multiple collaborators these days and wouldn't be helped by trying to treat individual collaborators as a level of trust.
2. We have examples of actors turning malicious https://snyk.io/blog/a-post-mortem-of-the-malicious-event-stream-backdoor/ . Updating during that good period would seem safe.
3. The trust provided per module would be a sum of all dependencies of that module, leading to huge amounts of auditing needed.
4. This could also be solved on the remote level by giving each trust level its own cert . This would likely mimic https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/ 's epic migration.

I still am unclear on the actual wins of signing a module itself. Especially if when dumped on disk it might not match what was signed due to scripts or fs mutation.

**Reply by wperron** [reply]
I agree with @bmeck here, I don't think these issues are new to URL modules, they are already present in the current ecosystem. Putting my security tin foil hat on for a minute, I'd argue it's best practice to use a lock-file with a hash of the source code downloaded, or to vendor dependencies with source code and to run static analysis on the built artifact.

I think an allow-list for specific domains and TLS verification + using an ETag header are sufficient - the rest can be addressed in userland

---

Should also point out that the exploit detailed in the linked article wouldn't have been possible without post-install scripts, which are a feature of the npm cli. URL imports wouldn't be able to support those anyway.

**Reply by mcollina** [reply]
> I agree with @bmeck here, I don't think these issues are new to URL modules, they are already present in the current ecosystem. Putting my security tin foil hat on for a minute, I'd argue it's best practice to use a lock-file with a hash of the source code downloaded, or to vendor dependencies with source code and to run static analysis on the built artifact.

Regarding https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610, my whole point was that URL modules where as susceptible as a central registry is, mainly due to the expiration of domain names. Calling this a known vulnerability is totally legit, at least folks could use countermeasures.

I don't think this is easy to mitigate in userland without some support from the runtime.

I _think_ we could all agree to have some level of certificate pinning (inside a lockfile) to mitigate this issue.

**Reply by ljharb** [reply]
Specifically, that attack was a vulnerability in *a trusted internal registry*, not in the public npm registry nor inherently in npm/registries, except in the inherent dangers of URLs that @mcollina outlines (which HTTPS imports increase the usage of).

**Comment by wperron** [comment]
I guess a good starting point would be to look at the npm cli because those problems are all also present there; they're just hidden behind the veneer of the cli.

I think we also need to clarify something from the discussion that happened over in the PR: when we say "HTTPS imports" do we mean that they are simply another way to reference and import modules into the node_modules, or are they implicitly dynamic imports?

**Reply by bmeck** [reply]
I'm not sure I understand. My understanding / expectation is it must be *possible* to do dynamic fetching of source texts over the network. I think trying to state it must be either always online or offline isn't really an easy topic since if we allow only one it invalidates some of the use cases of HTTPS imports.

**Reply by wperron** [reply]
that's how Deno does it and it's what makes more sense to me as well, I just wanted to clarify because a few of the comments on the PR seemed to assume it was _only_ dynamic imports.

**Comment by bmeck** [comment]
Concern to Mitigate: Cross Origin Access

E.G. `https:` to `file:`, `https:` to `node:`

**Reply by bmeck** [reply]
WHATWG already only allows only same domain from http/https schemes. I think the presumption is that `https:` could load `node:`, but I actually think that isn't really a priority as many/most modules don't actually need direct access to things like `fs:`. I do think in general that it is a security concern, but not a new attack vector. Without really strong reasons I do think requiring `https:` to have those kinds of `node:` modules dynamically injected as params or something seems fine to me.

**Reply by bmeck** [reply]
`node_modules` lives on `file:` so it would be prevented by nature of being cross origin and not allowing `https:` to `file:`.

**Reply by wperron** [reply]
Browsers have a very clear `origin` to compare against, how would that work from the server? For reference this is also an ongoing conversation in Deno https://github.com/denoland/deno/issues/4981

**Reply by bmeck** [reply]
@wperron same way as specced except we had to solve `file:` already for policies; for policies this was laid out with directory hierarchy as boundaries but same origin. I would note that due to how WHATWG's URL parser works `file:` resolution has an exploit if you don't realpath before performing checks.

**Reply by bmeck** [reply]
I would note that in general unlike the web we are not designing for a single base origin in node.

**Comment by mcollina** [comment]
HTTPS Agent support.

Given we are going to download _many_ files, hopefully from the same domain it's critical for "first time" startup performance to keep the connections alive.

**Reply by bmeck** [reply]
I don't think we should use Agent, it is mutable and means that an application could manipulate it to intercept things it shouldn't

**Reply by mcollina** [reply]
I think it should have an Agent which is kept private for esm.

**Reply by bmeck** [reply]
we could, but it cannot reuse the implementation currently in core

**Reply by targos** [reply]
Agent can be made safer to protect against user manipulation, right?

**Reply by bmeck** [reply]
@targos it could, but it wouldn't be simple or might even be infeasible given how many things manipulate it for instrumentation reasons

**Reply by bmeck** [reply]
If we populate the cache manually / as a non-standard setup does this still need to apply?

**Comment by mcollina** [comment]
Visible caching on disk.

I think we should cache what is downloaded in the project folder and recommend this to be committed to the repo itself. I propose we use a `web_modules` folder. This is also a mitigato to availability attacks.

**Reply by ExE-Boss** [reply]
Also, it might be a good idea to require a flag (or use an external utility) to enable remote fetching, which will be used to enable remote fetching and initialise the `web_modules` folder.

Without the flag, any `https://` import will only consult the local `web_modules` folder.

---

At the very least, there should be a way to disable remote lookup and only consult the local `web_modules` folder.

**Reply by wperron** [reply]
Visible caching on disk is a must, on that we agree, but why create yet another folder when reusing `node_modules/` would work just fine?

**Reply by mcollina** [reply]
> Visible caching on disk is a must, on that we agree, but why create yet another folder when reusing node_modules/ would work just fine?

The recommendation for `node_modules` is to not commit that to the repo. `web_modules` _should_ be committed to repo. Using two folders makes this simpler.

**Reply by mcollina** [reply]
> Also, it might be a good idea to require a flag (or use an external utility) to enable remote fetching, which will be used to enable remote fetching and initialise the `web_modules` folder.
> 
> Without the flag, any `https://` import will only consult the local `web_modules` folder.
> 
> At the very least, there should be a way to disable remote lookup and only consult the local `web_modules` folder.

I prefer a opt-in flag to download but also an opt-out flag for production usage is good. An env variable would also be good, as I think this is something that individual devs would like to change.

**Reply by wperron** [reply]
why should modules modules referenced via https:// be vendored locally but not modules referenced via their npm short identifier? at the end of the day, they'll all be pulled from a remote source, and content can be locked by the lock-file anyway.

**Reply by mcollina** [reply]
> why should modules modules referenced via https:// be vendored locally but not modules referenced via their npm short identifier?

npm guarantees provenance and availability - a module cannot be unpublished. An application without a central registry could be attacked by making a single weak spot in the chain unavailable. As an example, the fix npm did for the left-pad attack requires a central registry - this is what enables us to not commit `node_modules`.

**Reply by wperron** [reply]
I understand these concerns, and they are valid ones. However, I think these can (and should) be mitigated in userland. Having the complete url in a package name makes it super easy to lint, and even crawl transitive dependencies. IMHO, this is something that needs to be clear to people when they import from a URL, but ultimately I don't think the runtime should treat them any differently; that's something that can be done in userland, through the npm cli for example.

Also, to play devil's advocate here; npm modules have been and could be unpublished from the npm regsitry. The left-pad fiasco is in my opinion just another argument for allowing third-parties to host their own packages easily: If large organizations could simply host their packages under their own domain names, we wouldn't have to worry about naming conflicts on the npm registry.

**Reply by bmeck** [reply]
> npm guarantees provenance and availability - a module cannot be unpublished. An application without a central registry could be attacked by making a single weak spot in the chain unavailable. As an example, the fix npm did for the left-pad attack requires a central registry - this is what enables us to not commit node_modules.

npm does unpublish packages (for various reasons and sometimes without notification); Any remote source , including npm, has downtime concerns as seen with Google today. These mitigations are not unique to HTTPS.

**Reply by mcollina** [reply]
> I understand these concerns, and they are valid ones. However, I think these can (and should) be mitigated in userland. Having the complete url in a package name makes it super easy to lint, and even crawl transitive dependencies. IMHO, this is something that needs to be clear to people when they import from a URL, but ultimately I don't think the runtime should treat them any differently; that's something that can be done in userland, through the npm cli for example.

We are not in agreement. We currently ship with a solution that is safe and secure. Not shipping one will be a reduction in capability for what Node.js ships now.

> Also, to play devil's advocate here; npm modules have been and could be unpublished from the npm regsitry.

This is incorrect. npm packages cannot be unpublished anymore since left-pad.

**Reply by bmeck** [reply]
> This is incorrect. npm packages cannot be unpublished anymore since left-pad.

The npm team does unpublish things if you give requests with reasons such as security issues, licensing, etc. They also unpublish things without direct request if they find similar issues. The only thing that changes is that you cannot use the CLI to do so. This also is not true for most corporate npm proxies (which many/most companies do have for private package reasons).

**Reply by wperron** [reply]
> We currently ship with a solution that is safe and secure.

well, not really: Node only ships with a local name resolution algorithm. The actual fetching logic is delegated to the npm cli or yarn. If _those_ are compromised, so are the modules pulled into node_modules, and that has happened in the past. Adding an https:// module resolver to the core of Node would be a net addition to the current way of installing and and using dependencies.

**Reply by targos** [reply]
This is derailing a bit from the original discussion. To get back to the original topic: `node_modules/` is used by the Node.js module resolver. Using it as a cache for `https://` modules would change the behavior of the resolver.

**Reply by wperron** [reply]
@targos clearly the module resolver will need to be updated to account for https modules, making them first-class citizens just like modules identified by their npm identifier, so I don't think there's a conflict there, plus it makes the existing tooling and vendoring patterns just work out of the box, or at least makes it close enough that adapting it to https modules would be a relatively straight-forward upgrade

**Reply by bmeck** [reply]
@targos yes/no on changing the behavior, there is precedent on using it for other things like `.cache`, `.bin`, etc. Adding support for things that cannot be installed via a registry due to the naming conventions enforced by the registry like these wouldn't have any real effect.

**Comment by mcollina** [comment]
HTTP Proxy support for downloading modules.

A good way to mitigate many availability attack is to download the modules through an HTTP proxy. Currently Node.js does not have this functionality built-in.

**Reply by bmeck** [reply]
Currently policies allow rewriting URLs, would allowing wildcard URL rewriting like export templates be enough?

**Reply by bmeck** [reply]
Also, can this be done as a follow on during experimentation?

**Comment by bmeck** [comment]
I've made https://github.com/bmeck/local-fs-https-imports to play around with the purely ahead of time idea before trying to merge such a feature into core itself.

**Reply by wperron** [reply]
I like the `node_modules/.https/` - seems like a nice way to make the old and the new play together nicely

**Comment by jimmywarting** [comment]
Concern To Mitigate: process

- process.exit()
- process.env

the hole process contains way to much information that can be leaked

**Reply by bmeck** [reply]
Unfortunately we can't put this in scope until we have some newer JS features; this likely has to wait for Realms or some such feature. We looked into policies and other mechanisms around limiting it for ESM in a few things in the past like: https://github.com/nodejs/node/pull/26334. We need a way to introduce new global scopes w/o creating new intrinsics for things like Array which we don't have a way to do so currently. Users could replace `globalThis.process` with a getter that does a callsite check like in that and the other PR we attempted to get through, but the overall performance hit is not negligible. Other ways to limit it for now would be up to source transforms detecting usage of the globals and replacing it with a corresponding import/require.

**Reply by bmeck** [reply]
adding: fetch() is coming and is also a major concern

**Comment by bmeck** [comment]
I've made a doodle to have a call to discuss how to move this forward, likely this call will try to recur every 2 months as various things are likely to be around phasing and when/how to address concerns in this discussion when implementing this and how to move towards a stable release.

https://doodle.com/poll/wsdihf3q9gqtyack

**Reply by mcollina** [reply]
Can we move it one week further? I'm off that week.

**Reply by bmeck** [reply]
bumped, updated to new link

**Reply by mcollina** [reply]
Filled

**Reply by bmeck** [reply]
Given the replies Thursday June 10th 1PM-2PM UTC seems to work and I will get a calendar entry added. Likely this one is just about phasing as mentioned above not about implementation details / shipping.

**Reply by MylesBorins** [reply]
I'd love to join!

**Reply by bmeck** [reply]
Thanks to @MylesBorins for making the [calendar event]( https://www.google.com/calendar/event?eid=Nmxxam9lbGcyZjFyZHVwMmVidmRvMGx2c2sgbm9kZWpzLm9yZ19ucjc3YW1hOHA3ZDdmOWFqcnBudTUwNmM5OEBn ) on the [Node.js public calendar]( https://calendar.google.com/calendar/embed?src=nodejs.org_nr77ama8p7d7f9ajrpnu506c98%40group.calendar.google.com ). See you all in a day and a bit!

**Comment by guybedford** [comment]
Concern: npm packages using HTTPS imports.

Posting this to link discussion from @ljharb in https://github.com/nodejs/node/pull/36328#issuecomment-860359362 re implications for supporting in node_modules.

Personally I would be concerned about npm packages being published that use HTTPS imports as that very much changes the security model / contract that npm provides to users in being able to comprehensively replicate a module graph via only a node_modules filesystem model.

This likely won't be a problem initially as no one would publish a package to npm that isn't supported in older Node.js versions, but once support reaches all versions, some company might think it's a good way to automatically manage updates / ensure registered code delivery.

While that is already possible today via network based processes and even install hooks, it's a pattern that we have luckily so far avoided from happening too much and I hope we can continue to work towards ensuring strong guarantees around the registry and package contracts that ensure end user agency.

I haven't watched the video so apologies if this is repeating talking points already said, feel free to fill in further here.

**Reply by bmeck** [reply]
Per discussions above and video, network access would not be enabled by default when running node. During experimental it will be enabled automatically under the flag while offline cache is worked out since the flag is the opt-in.

**Reply by guybedford** [reply]
Thanks, not all of us have time to watch videos, and well-maintained high-level summaries would help a lot in ensuring adequate onboarding for those familiarizing.

Is the flag a CLI flag with the intention to unflag at some point? Is this not something that can be integrated into policy?

**Reply by bmeck** [reply]
> Is the flag a CLI flag with the intention to unflag at some point? Is this not something that can be integrated into policy?

Various features were punted and behavior under the flag will change to accomodate them; the flag as it exists likely won't be unflagged until policy integration etc. is done but even with policy integration features like manipulating how redirects are expected to work for HTTP are not possible currently. We do have a POC at https://github.com/bmeck/local-fs-https-imports but it doesn't properly mimic various nuances of web based loading.

**Reply by targos** [reply]
It's a CLI flag (`--experimental-https-modules`) and I don't think we discussed whether it would be unflagged at some point.

**Reply by bmeck** [reply]
@targos we did mention in the docs that there was work for unflagging and post unflagging per meeting notes. I might be misunderstanding this question?

**Reply by guybedford** [reply]
I want to be clear that I would personally be against any PR that allows unflagged HTTPS imports without a mandatory policy configuration in place. The flagged implementation as a starting point seems great though and I am pleased to see progress here.

**Reply by bmeck** [reply]
@guybedford are you against using the HREFs that contain `https:` or against the network access? Network access has a lot of config bits that need ironing out in these discussions so if there is a specific concern on network access just do a quick search and/or just make a new thread about the issue. To my knowledge no one on the call was demanding that we support network access unflagged without some kind of opt-in mechanism.

**Reply by guybedford** [reply]
@bmeck `https:` imports since when included in packages published to npm they fundamentally change the npm registry from something immutable into something mutable.

**Reply by bmeck** [reply]
@guybedford so, if I am understanding, even with the opt-in you don't want to allow those string specifiers?

**Reply by targos** [reply]
@bmeck 

> we did mention in the docs that there was work for unflagging and post unflagging per meeting notes. I might be misunderstanding this question?

My bad, I forgot we separated unflagging and experimental status.

**Reply by guybedford** [reply]
I'm more than happy with these specifiers, and as I say I'm pleased to see progress on this, but the restriction that I think is important is that the **end user** explicily allows this behaviour either via a flag or an explicit policy to enable it, instead of not knowing if one of their package dependencies happens to be using https imports.

**Reply by targos** [reply]
> the end user explicily allows this behaviour either via a flag or an explicit policy to enable it, instead of not knowing if one of their package dependencies happens to be using https imports.

And I suppose you mean that this flag should be specific to https imports in `node_modules`, i.e., it should be possible to use https imports in application code while disallowing it in dependencies?

**Reply by bmeck** [reply]
@guybedford got it. split that specific concern to: https://github.com/nodejs/node/discussions/36430#discussioncomment-868665

**Reply by guybedford** [reply]
@bmeck one thing that might make this easier is if Node.js were to have a "default policy file" check without the policy flag being necessary - so that the entire workflow can behave unflagged in the end, but like CORS or any other web security measure just requires an initial setup.

**Reply by bmeck** [reply]
@guybedford I think policy discussions could be done elsewhere.

**Reply by guybedford** [reply]
Certainly - my point is just that this doesn't stop there ultimately being an entirely "flagless" scenario for the use case.

**Reply by bmeck** [reply]
@guybedford does this actually expand to Buffer/process as well?

**Comment by bmeck** [comment]
meta concern to mitigate: accidental inclusion of transitive mutable state via node_modules.

A user should not have node_modules introduce transitive mutable state unless they opt-in to allow such behavior. This is important to ensure that node_modules can be seen as effectively immutable. While node_modules are not technically immutable today the general practice allows them to be treated as immutable; the introduction of network based transitive mutable state is actively more unstable and a concern around security and robustness that needs to be mitigated by other means as seen in this discussion.

**Reply by boneskull** [reply]
@bmeck can you please provide an example of network-based transitive mutable state?

**Reply by bmeck** [reply]
@boneskull a good example is any JS CDN that redirects to the "latest" version of a package E.G.

* https://esm.sh/uuid => https://cdn.esm.sh/v66/uuid@8.3.2/es2021/uuid.js does this via in source re-exporting
* https://unpkg.com/uuid?module => https://unpkg.com/uuid@8.3.2/dist/esm-node/index.js?module does this via a redirect

Then uuid gets updated so now they magically get updated to point to 8.3.3 for example without explicit updating on behalf of the person running the application.

**Reply by ljharb** [reply]
Also, any time a domain name's ownership changes, the new owner could serve whatever they want from the site.

**Reply by wperron** [reply]
That's also been a concern of mine, I wonder if it'd possible/advisable to add whois info in the lockfile for https deps 🤔

**Reply by bmeck** [reply]
@wperron Just about any out of band info can be added, time it is synchronized and checked is the main question. It is a security vector for example to do a WHOIS query.

**Reply by guybedford** [reply]
@bmeck do you have tooling for integrity lockdown that can work with remote URLs and policy?

**Reply by bmeck** [reply]
@guybedford node-policy needs to be updated, so no not at this time.

**Comment by bmeck** [comment]
Concern to mitigate: using other protocols to hop to local resources.

It is possible currently for `data:` to access `node:` and `file:` URLs (of which there is some usage in the wild). Since it is possible at least in this PR to load `data:` via `http{s,}:` we might want to ban loading these by default.

**Reply by bmeck** [reply]
I've banned everything except network dependencies in the PR for now.

**Reply by targos** [reply]
Note that depending on how the main application is started (notably via a CommonJS entry point), it's still possible to access the `node:` scheme using `globalThis.process.mainModule.require()`.

**Reply by bmeck** [reply]
@targos yup, this is known and we tried to fix `Buffer`/`process` so they were not implicit globals in ESM but failed to do so due to a perf concern. So... ambient ways to mitigate it are left up to users and not covered by things like policies. A lot of mitigations are about limiting how things can be accessed so they can properly be mitigated, and unfortunately not mitigating them in core itself. Removing `data:` for example means that you have to grab them via globals, or be passed them somehow via dependency injection. It removes `import` from the problem areas.

**Comment by bmeck** [comment]
Concern to mitigate: preventing injection of arbitrary locations to `import()`

It should be possible to differentiate if a dependency is allowed to be loaded by dynamic import. Static import has hard coded strings and is less vulnerable to manipulation than `import()` that can take an arbitrary expression. Unfortunately due to JS spec we cannot prevent `import()` from resolving something that was already loaded by static `import`.

CC: @vdeturckheim per https://docs.google.com/document/d/1tS80LmzraV4yXEYIQ-fWaCm2MIpXfbtrSYjsI8kWKko/edit#heading=h.ifctpvwuev0a

**Reply by bmeck** [reply]
CC: @jasnell from same doc (woops)

**Reply by targos** [reply]
How does the spec prevent us from intercepting the call before it hits the cache?

**Reply by bmeck** [reply]
@targos it states that `import()` and `import` share the cache and deterministically share the result (this enforced by V8 source text module local cache, not the global cache which node controls). If one succeeds, the other must also succeed in perpetuity for the same source text and cache key (URL + import assertions these days [may grow]). So if you do `import "fs";` dynamic import feeds through to FinishDynamicImport which feeds into the same mechanism as static imports (HostResolveImportedModule) which has the restriction: "Each time this operation is called with a specific referencingScriptOrModule, specifier pair as arguments it must return the same Module Record instance if it completes normally."

**Reply by targos** [reply]
Doesn't "if it completes normally" allow us to throw/reject the promise?

**Reply by bmeck** [reply]
@targos that just means after the first success it always succeeds, so you can only error out if it didn't succeed before.

**Reply by guybedford** [reply]
Reposting - Whether it is static or dynamic, could this not be handled via standard policy rules for allowed imports?

> currently policy doesn't distinguish dynamic vs static. we could punt it to policies but even then spec doesn't let us completely prevent https: import from dynamic import if it was already loaded via static import.

@bmeck what I mean is that rather than specially handling dynamic imports as a policy, that the policy of just restricting which packages can import which modules can work equally well for static and dynamic imports.

**Reply by bmeck** [reply]
i think that is fine, just trying to be sure while i was reviewing the doc above

**Comment by guybedford** [comment]
Whether it is static or dynamic, could this not be handled via standard
policy rules for allowed imports?

**Reply by bmeck** [reply]
this didn't thread properly I think. currently policy doesn't distinguish dynamic vs static. we could punt it to policies but even then spec doesn't let us completely prevent `https:` import from dynamic import if it was already loaded via static import.

**Comment by sosoba** [comment]
IMHO this feature should be provided by `--loader` plugin according to the requirements of the situation. See https://nodejs.org/dist/latest-v16.x/docs/api/esm.html#esm_https_loader .

Therefore, it is more important to work on loaders API stability which will allow to loading modules from anywhere (e.g. from tgz file etc.) and however (on the fly, with file cache etc.).

**Reply by JakobJingleheimer** [reply]
There's quite a lot of rationale behind supporting this in core. It is possible to do with a loader, but that would likely result in divergent behaviour, security vulnerabilities, etc.
